src/centralserver/admin-service/src/main/java/org/niis/xroad/centralserver/restapi/service/InitializationService.java

```diff
@@ -0,0 +1,282 @@
+        }
+        if (!isSWTokenInitialized && configDto.getSoftwareTokenPin().isEmpty()) {
+            errorMetadata.add(ERROR_METADATA_PIN_NOT_PROVIDED);
+        }
+        if (isServerAddressInitialized && !configDto.getCentralServerAddress().isEmpty()) {
+            errorMetadata.add(ERROR_METADATA_SERVER_ADDRESS_EXISTS);
+        }
+        if (!isServerAddressInitialized && configDto.getCentralServerAddress().isEmpty()) {
+            errorMetadata.add(ERROR_METADATA_SERVER_ADDRESS_NOT_PROVIDED);
+        }
+        if (isInstanceIdentifierInitialized && !configDto.getInstanceIdentifier().isEmpty()) {
+            errorMetadata.add(ERROR_METADATA_INSTANCE_IDENTIFIER_EXISTS);
+        }
+        if (!isInstanceIdentifierInitialized && configDto.getInstanceIdentifier().isEmpty()) {
+            errorMetadata.add(ERROR_METADATA_INSTANCE_IDENTIFIER_NOT_PROVIDED);
+        }
+        if (!errorMetadata.isEmpty()) {
+            log.debug("collected errors {}", errorMetadata.stream().reduce(
+                    (String result, String toAdd) -> result.concat(", " + toAdd)
+            ).orElse("none"));
+            throw new InvalidInitParamsException("Empty, missing or redundant parameters provided for initialization",
+                    errorMetadata);
+        }
+    }
+
+    private void initializeCsSystemParameters() {
+        systemParameterService.updateOrCreateParameter(
+                CONF_SIGN_DIGEST_ALGO_ID,
+                DEFAULT_CONF_SIGN_DIGEST_ALGO_ID
+        );
+        systemParameterService.updateOrCreateParameter(
+                CONF_HASH_ALGO_URI,
+                DEFAULT_CONF_HASH_ALGO_URI
+        );
+        systemParameterService.updateOrCreateParameter(
+                CONF_SIGN_CERT_HASH_ALGO_URI,
+                DEFAULT_CONF_HASH_ALGO_URI
+        );
+        systemParameterService.updateOrCreateParameter(
+                SECURITY_SERVER_OWNERS_GROUP,
+                DEFAULT_SECURITY_SERVER_OWNERS_GROUP
+        );
+    }
+
+    private void initializeGlobalGroupForSecurityServerOwners() {
+        Optional<GlobalGroup> securityServerOwnersGlobalGroup = globalGroupRepository
+                .getByGroupCode(DEFAULT_SECURITY_SERVER_OWNERS_GROUP);
+        if (securityServerOwnersGlobalGroup.isEmpty()) {
+            GlobalGroup defaultSsOwnersGlobalGroup = new GlobalGroup();
+            defaultSsOwnersGlobalGroup.setGroupCode(DEFAULT_SECURITY_SERVER_OWNERS_GROUP);
+            securityServerOwnersGlobalGroup = Optional.of(defaultSsOwnersGlobalGroup);
+        }
+        securityServerOwnersGlobalGroup.get().setDescription(DEFAULT_SECURITY_SERVER_OWNERS_GROUP_DESC);
+        globalGroupRepository.save(securityServerOwnersGlobalGroup.get());
+    }
+
+    private boolean isSWTokenInitialized() {
+        boolean isSWTokenInitialized = false;
+        List<TokenInfo> tokenInfos;
+        try {
+            tokenInfos = signerProxyService.getTokens();
```

Comment:
Instead of fetching all tokens, one could just fetch the one with SSL_TOKEN_ID (SignerProxy.getToken(id)). Simplifies the implementation.

Reply:
Fixed in: 
c0772c7